addOns/dns/src/main/java/org/zaproxy/addon/dns/SpfScanRule.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.dns;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.commons.httpclient.URIException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.AbstractHostPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.addon.dns.exceptions.TooManyRecordsException;
+
+public class SpfScanRule extends AbstractHostPlugin {
+    private static final int id = 90040;
+    private static final Logger LOGGER = LogManager.getLogger(SpfScanRule.class);
+    private static final String MESSAGE_PREFIX = "dns.";
+    private volatile boolean enabled = true;
+    private static List<String> reviewedDomains = new ArrayList<String>();
```

Comment:
- This should take into account the session changes (e.g. a session listener added by the extension which resets the state on new session and session changed).
- Should handle concurrency, multiple scan rules can be executed at the same time.

Reply:
To handle concurrency, I've changed the `ArrayList` into `Vector`. I expect that to be enough.

For the first part, I've tried to implement the `SessionChangedListener` interface on `SpfScanRule`, but I do not know how to register it into the hook (or even if it is required). Could you suggest me a way?

Reply:
`Collections.synchronizedList(new ArrayList<>());` would be more clear that synchronization is wanted.

Better not implement directly on the scan rule, there will be multiple instances. Add a nested class in the extension which calls a static method of the scan rule (see `hook` method in the extension).